### PR TITLE
WT-2725 Change how we detect whether a page can be evicted when cache…

### PR DESCRIPTION
**Pull request by daveh86**
… is full

### CODE REVIEW

File: src/evict/evict_lru.c [hunk 1]
```diff
@@ -1273,8 +1262,7 @@ __evict_walk_file(WT_SESSION_IMPL *session, u_int *slotp, uint32_t flags)
 		 * running last time we wrote the page has since rolled back,
 		 * or we can help get the checkpoint completed sooner.
 		 */
-		if (modified && !LF_ISSET(
-		    WT_EVICT_PASS_AGGRESSIVE | WT_EVICT_PASS_WOULD_BLOCK) &&
+		if (modified &&
 		    (mod->disk_snap_min == S2C(session)->txn_global.oldest_id ||
```
Comment (michaelcahill) [inline src/evict/evict_lru.c]: This part of the change should be unrelated.  Basically, this check is an optimization -- don't try to evict pages if we can tell with these quick checks that eviction will fail.

File: src/include/btree.i [hunk 2]
```diff
@@ -1083,6 +1083,13 @@ __wt_page_can_evict(WT_SESSION_IMPL *session,
 	/* Pages that have never been modified can always be evicted. */
 	if (mod == NULL)
 		return (true);
+	 /*
+	  * If the page is clean but has modifications that appear too
+	  * new to evict, skip it.
+	  */
+	if (!__wt_page_is_modified(page) &&
+	    !__wt_txn_visible_all(session, mod->rec_max_txn))
+	        return (false);
```
Comment (michaelcahill) [inline src/include/btree.i]: If you change this to `if (!__wt_txn_visible_all(session, mod->rec_max_txn)) ...`, is that sufficient to avoid the assertion (without the unrelated change to `evict_lru.c`)?

**Comment by daveh86** [comment]
test this please

File: src/include/btree.i (hunk 2 above)
Comment (daveh86) [inline src/include/btree.i]: I've patched these changes in for testing and adding the check for the WT_EVICT_PASS_WOULD_BLOCK allows the test to pass, however adding the WT_EVICT_PASS_AGGRESSIVE flag as an exclusion causes the test to fail.

**Comment by daveh86** [comment]
@agorrod , would you mind taking a look at this?

File: src/include/btree.i [hunk 3]
```diff
@@ -1085,6 +1085,14 @@ __wt_page_can_evict(WT_SESSION_IMPL *session,
 		return (true);
 
 	/*
+	 * If the page is clean but has modifications that appear too
+	 * new to evict, skip it.
+	 */
+	if (!__wt_txn_visible_all(session, mod->rec_max_txn) ||
+	    !__wt_txn_visible_all(session, mod->update_txn))
```
Comment (agorrod) [inline src/include/btree.i]: @daveh86 I suggested looking for different transaction IDs that might be more suitable here, but I don't think ``mod->update_txn`` is the one we are interested in. The ``update_txn`` is used as an advisory value, and shouldn't be involved in correctness checks (which we are pursuing here).

The expectation in the code is that either:
* The page is clean and ``rec_max_txn`` accurately reflects the ID of the most recent update on the page (which needs to be kept around as long as the oldest reader was started before the update).
* The page is dirty, and the state of the update lists are checked during reconciliation.

Can you reproduce this failure reliably? If so, could you capture the assertion in a debugger? We would want to be able to see the fully populated page-modify structure, so we may need to restructure the code a bit to find the problem before cleaning up content.

**Comment by agorrod** [comment]
Closing. This will be replaced with a different version.
